orttraining/orttraining/python/training/utils/hooks/_subscriber_manager.py

```diff
@@ -0,0 +1,369 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+
+from collections import abc
+from typing import Callable, List, Union
+
+import warnings
+import torch
+
+from onnxruntime.training.ortmodule import ORTModule
+
+
+class ModuleHookSubscriberBase:
+    """
+    Base class for all module hook subscribers.
+    Currently the hook here only means post forward hook and pre backward hook.
+
+    A module hook subscriber is a class that implements the forward and backward function.
+    > The forward hook is called after the activation is generated in the forward path.
+    > The backward hook is called before the activation gradient is computed.
+
+    Forward path:
+        Module_A generate activation tensor_a --> Post forward hook (calling subscribers's forward one by one) -->
+        Module_B generate activation tensor_b --> ...
+
+    Backward path:
+        Module_B backward run, tensor_b's gradient is computed as tensor_b_grad -->
+        Pre backward hook (calling subscribers's backward one by one) -->
+        Module_A backward run, tensor_a's gradient is computed as tensor_a_grad
+
+    Be noted: the "Pre"/"Post" is described from the perspective of the Module_A.
+    """
+
+    def __init__(self, start_step: Union[None, int], end_step: Union[None, int]):
```

Comment:
When would start step/end step be none? Would that imply that the subscriber is disabled? 
I think start step should be only an int and defaulted to 0. End step can be None (meaning run indefinitely). What do you think?

So, we probably can update the signature to be

```py
def __init__(self, start_step: int = 0, end_step: Optional[int] = None):
```